File: backend/infraction-service/app/com/radovan/play/providers/NatsUtilsProvider.scala
package com.radovan.play.providers

import com.google.inject.Provider
import com.radovan.play.utils.NatsUtils
import jakarta.inject.Inject
import play.api.inject.ApplicationLifecycle

import scala.concurrent.Future

class NatsUtilsProvider @Inject()(lifecycle: ApplicationLifecycle) extends Provider[NatsUtils] {

  private val natsUtils = new NatsUtils
  natsUtils.init()


  lifecycle.addStopHook { () =>
    Future.successful {
      natsUtils.closeConnection()
    }
  }

  override def get(): NatsUtils = natsUtils
}

File: backend/registration-service/app/com/radovan/play/providers/NatsUtilsProvider.java
package com.radovan.play.providers;

import com.google.inject.Provider;
import com.radovan.play.utils.NatsUtils;
import jakarta.inject.Inject;
import play.inject.ApplicationLifecycle;

import java.util.concurrent.CompletableFuture;

public class NatsUtilsProvider implements Provider<NatsUtils> {

    private final NatsUtils natsUtils;

    @Inject
    public NatsUtilsProvider(ApplicationLifecycle lifecycle) {
        this.natsUtils = new NatsUtils();
        this.natsUtils.init();

        // registrujemo shutdown hook
        lifecycle.addStopHook(() -> {
            natsUtils.closeConnection();
            return CompletableFuture.completedFuture(null);
        });
    }

    @Override
    public NatsUtils get() {
        return natsUtils;
    }
}

